### ISSUE TITLE
Reduce the generated code size when using Fragments

### ISSUE THREAD

**Issue by snijsure** [apollographql/apollo-kotlin#5513]
### Question

Are there any recommendations that will help in reducing the size of classes that are generated by apollo-android.

In our code base we use fragments extensively. We see there are ton of "Impl_ResponseAdapter" classes generated for fragments wherever the specific fragments are referenced. That seems to "bloat" the overall size. 

I am using apollo version 3.7.5

Would appreciate any tips to reduce size of generated classes.

**Comment by martinbonnin** [comment]
There's no silver bullet unfortunately. The `Impl_ResponseAdapter` are the JSON parsers. They're needed to parse the response into typesafe models and last time I checked, the bytecode was in the similar range as other alternative JSON parsers. 

The one thing that comes to mind is disabling data classes with `generateMethods`:

```kotlin
apollo {
  service("api") {
    generateMethods.set(emptyList())
  }
}
```


See https://github.com/apollographql/apollo-kotlin/pull/5212 for more details.

Other thing that comes to mind is make sure you're shrinking your code with R8 or proguard.

Besides that, we're trying quite hard not to generate unused code. If you can share an APK Analyzer (In Android Studio: Build -> Analyze APK), we'll look into it but saving more bytecode will probably be challenging without dropping some typesafety.

**Comment by snijsure** [comment]
Thanks @martinbonnin as always for suggestion.

Yes we do use fullR8mode for app.

Will explore proper use of generateMethods option.

**Comment by martinbonnin** [comment]
@snijsure Something else I just remembered if you want to shave on the bytecode size, another interesting thing would be to use [persisted queries](https://www.apollographql.com/docs/kotlin/advanced/persisted-queries/). Using persisted queries is also more secure and saves network bandwidth so it really has mostly upsides at the cost of some build setup.

Using `generateQueryDocument.set(false)` you can remove the queries from your app completely. You can read more about persisted queries [here](https://www.apollographql.com/docs/kotlin/advanced/persisted-queries/)

**Comment by snijsure** [comment]
We do use persisted queries in our app, I will explore the ```generateQueryDocument.set(false) ``` option as well.
I see all of these options are available in Apollo 4.0.X which is currently in beta-4, is there a "time line" in mind when 4.0 is likely to become officia? 

In the meantime I am actively migrating our app from 3.x to 4.x-beta for testing.

**Comment by martinbonnin** [comment]
You can get latest estimates in the [ROADMAP](https://github.com/apollographql/apollo-kotlin/blob/main/ROADMAP.md).

4.0 will build with [K2](https://blog.jetbrains.com/kotlin/2023/02/k2-kotlin-2-0/) so it'll require a stable version of K2 but overall the goal is early 2024.

PS: `generateQueryDocument` is in 3.x as well (but not `generateMethods`)
PS2: in terms of stability, 4.x is pretty stable. The API is not but if you're building in app (and not a lib), using it should be pretty safe.
